### ISSUE TITLE
Ability to apply ExtensionMethod project wide

### ISSUE THREAD

**Issue by aikar** [rzwitserloot/lombok#922]
I asked about this before here: https://groups.google.com/forum/#!topic/project-lombok/4NGp9Rzr6aM
Figured we should have an open issue for the request.

Requesting ability to enable ExtensionMethod across an entire folder using the lombok.config file, to be able to apply it to an entire package and configure it.

Would greatly improve the use of this feature as I can not realistically see us adding the annotation to every class created, but would love to be able to extend API's we are using.

**Comment by rspilker** [comment]
Agreed. This was the whole inspiration for the [lombok configuration system](https://projectlombok.org/features/configuration.html). We're still working on it.

**Comment by aikar** [comment]
Any update on where this lands priority wise? I really believe this would be lombok's #1 feature once this is done.

**Comment by sizebay** [comment]
+1

I agree with @aikar 
> I really believe this would be lombok's #1 feature once this is done.

**Comment by aikar** [comment]
bumping this -- @rspilker any chance this could please make a near priority?

**Comment by aikar** [comment]
@rspilker @rzwitserloot any chance of this coming to reality soon?

**Comment by jasonCodesAway** [comment]
bump!

**Comment by miere** [comment]
+1

> I really believe this would be lombok's #1 feature once this is done.

I agree with @aikar as well

**Comment by miere** [comment]
I can imagine how expensive would it be to check for "unknown" method that might be extensions. Specially if we're walking through all classes in a given directory (and sub-directories). I wonder if changing the mechanics on how we define extension methods would relief the burden it imposes and make it easier to be implemented.

**Comment by Maaartinus** [comment]
@miere 
> I can imagine how expensive would it be to check for "unknown" method that might be extensions. Specially if we're walking through all classes in a given directory (and sub-directories). I wonder if changing the mechanics on how we define extension methods would relief the burden it imposes and make it easier to be implemented.

I don't know. With extensions defined project-wide, I guess that scanning the extension classes only once per compilation should be possible. Thereafter, two `HashMap` lookups should do:

- the first hash map lookup by name returns a map to be used in the next step
- the second hash map lookup by receiver type returns the class defining the extension method

This goes as follows:

- The first step eliminates all methods not named as an extension method. This is a sure and simple speed up.
- The second step uses the receiver type directly. When there's no value for it, a test like in [this line](https://github.com/rzwitserloot/lombok/blob/master/src/core/lombok/javac/handlers/HandleExtensionMethod.java#L186) gets done and the result gets stored back. When no extension method applies, a sentinel value gets used.

I may be wrong with it all as I know too little about the compiler internals. If I'm right, it could be way faster than what gets done now and that independently of the number of extension methods.

@rzwitserloot Could you comment on it? I guess, the idea of storing all method names should apply to the current code, too. Something like
```
/* Set<String> */ extensionNames = extensions.stream()
    .flatMap(e -> e.extensionMethods.stream().map(m -> m.name))
    .collect(Collectors.toSet());
```
in the `ExtensionMethodReplaceVisitor` constructor and a check like
```
if (!extensionNames.contains(methodName)) return;
```
in `handleMethodCall` could save most of the effort for a majority of methods, couldn't it?

**Comment by rzwitserloot** [comment]
It'll have a significant impact.

We are focussing on different high-impact features right now; see the ones marked as 'next milestone'.

**Comment by Rawi01** [comment]
I prototyped this for javac to check out the performance impact during compilation. It seems like it takes 10%-20% more time to complete compilation. I also added an additional name and parameter count check to skip reslution, without these it is really slow.

**Comment by chaganiu** [comment]
bump.  any updates on this.  Even a partial solution like being able to inherit `@ExtensionMethod` from a base class would help.

**Comment by rzwitserloot** [comment]
No updates on this; for headline features, meta annotations is currently ahead in the queue, and ExtensionMethod needs a whole bunch of love in the first place.

There's also the topic of generics. On the one hand, doing something like this is amazing:

```java
public static String join(List<? extends CharSequence> listOfStrings, String separator) { ... }
```

That would mean any `List<String>` or `List<StringBuilder>` or whatnot just has a `join(', ')` method, but other lists simply don't have it.

On the other hand it complicates matters.

Keeping the issue open because whilst it's got a very long list of caveats (really high costs), the potential benefits of this are similarly sky high.
